sections/advanced/email-testing.md

```diff
@@ -0,0 +1,213 @@
+
+
+```javascript
+// cypress/plugins/index.js
+
+const task = require('some-plugin/task');
+const percyHealthCheck = require('@percy/cypress/task'); // or any other plugin you may need
+const mailosaurTasks = require('./mailosaur-tasks'); // our mailosaur module
+
+// This is a pattern to merge all Cypress tasks
+const all = Object.assign({},
+  percyHealthCheck,
+  task,
+  mailosaurTasks
+)
+
+module.exports = (on, config) => {
+  on("task", all);
+}
+
+////////
+
+// cypress/plugins/mailosaur-tasks.js (this could be anywhere)
+
+// the npm package
+const MailosaurClient = require('mailosaur');
+// we used a static file for envVars. cypress.env.json file can cause issues in CI 
+// There can be other solutions, do your best here.
+const envVars = require('../../cypress.json');
+const mailosaurClient = new MailosaurClient(envVars.MAILOSAUR_API_KEY);
+
+// replicate Mailosaur's npm code from api docs
+// https://docs.mailosaur.com/docs/fetching-messages
+/** finds the most recent email message to the given email*/
+const findEmailToUser = async (userEmail) => {
+  let message = await mailosaurClient.messages.get(envVars.MAILOSAUR_SERVERID, {
+    sentTo: userEmail
+  }, { timeout: 25000 }); // time to wait for an email to arrive
+  return message;
+};
+
+// other useful utilities can include the below. You can replicate them using the api docs.
+
+// checkServerName()
+// createEmail()
+// deleteAMessage(messageId)
+// listAllMessages()
+
+module.exports = { checkServerName, createEmail, findEmailToUser, listAllMessages, deleteAMessage };
+
+```
+### Other Useful helper functions that Mailosaur npm package does not provide *(as far as we know)*
+
+We can harmonize Rest api / `cy.request()` approach with npm package / `cy.task()` to build our own utility.
+
+```javascript
+/** Given user email, returns the id of the email to that user. Good example of hybrid utility functions */
+const getEmailId = email => cy.task('findEmailToUser', email).its('id');
+
+/** Deletes 1 email message by message id. Can be useful if you want to delete the message after running the test. */
+const deleteEmailById = id => {
+  return cy.request({
+    method: 'DELETE',
+    url: `${Cypress.env('MAILOSAUR_API')}/messages/${id}`,
+    headers: { // important detail
+      authorization: Cypress.env('MAILOSAUR_PASSWORD')
+    },
+    auth: { // important detail
+        user: Cypress.env('MAILOSAUR_API_KEY'),
+        password: '' // any pw or empty pw will do
+    },
+    retryOnStatusCodeFailure: true // because we can
+  });
+};
+
+/** Deletes the most recent email sent to the user. Useful for resetting state. */
+export const deleteEmail = email =>
+  getEmailId(email)
+    .then(id => deleteEmailById(id));
+
+```
+
```

Comment:
I’m also using Mailosaur https://mailosaur.com/api/messages/await api, which is super useful, but I see is no longer in documentation 😱 I hope they don’t mean to deprecate it, it’s basically 80% of what makes mailosaur useful. Anyway an example of the await api might fit in here.

Reply:
they do not have it in the docs anymore. What does it do? Can you share the code?

There may be a proposed alternative.

Reply:
Sorry, was busy.  Simple code would look something like this:

```
cy
  .request({
      method: 'POST',
      url: `https://mailosaur.com/api/messages/await?server=${Cypress.env('mailosaurServer')}`,
      headers: {
        'authorization': Cypress.env('mailosaurAuth')
      },
      body: 
        {
          'sentTo': email,
          'subject': subject
        }
    })
```
and what the request would do is, that it would give a response after the message is found. In real life, it will take a couple of seconds fo the email to be delivered, and this method takes this into account. if it is not delivered in 30 seconds, request will return 204. if it does, it returns the message details immediately. to me, it fits perfectly into Cypress logic.

Reply:
I see. This is a utility to get the message sent to a certain email.
They do not  provide this conveniently in the api. 

This is why we had to build a helper function like this:

```javascript
// as a cy.task() util

// https://docs.mailosaur.com/docs/fetching-messages more options here
/** finds the most recent email message */
const findEmailToUser = async (userEmail) => {
  let message = await mailosaurClient.messages.get(envVars.MAILOSAUR_SERVERID, {
    sentTo: userEmail
  }, { timeout: 25000 });
  return message;
};

// as a helper file
/** Given user email, returns the id of the email to that user */
const getEmailId = email => cy.task('findEmailToUser', email).its('id');

````